File: backend/app/models/return_request.py
from .. import db
from datetime import datetime

class ReturnRequest(db.Model):
    __tablename__ = 'return_requests'
    
    id = db.Column(db.Integer, primary_key=True)
    reseller_id = db.Column(db.Integer, db.ForeignKey('resellers.id'), nullable=False)
    product_id = db.Column(db.Integer, db.ForeignKey('products.id'), nullable=False)
    order_id = db.Column(db.Integer, db.ForeignKey('order_requests.id'), nullable=False)
    quantity = db.Column(db.Integer, nullable=False)
    reason = db.Column(db.Text, nullable=False)
    status = db.Column(db.String(20), default='pending')  # pending, approved, rejected, processed
    request_date = db.Column(db.DateTime, default=datetime.utcnow)
    processed_date = db.Column(db.DateTime)
    created_at = db.Column(db.DateTime, server_default=db.func.now())
    updated_at = db.Column(db.DateTime, server_default=db.func.now(), onupdate=db.func.now())
    
    def __repr__(self):
        return f'<ReturnRequest {self.id} by Reseller {self.reseller_id}>'

    def to_dict(self):
        return {
            'id': self.id,
            'reseller_id': self.reseller_id,
            'product_id': self.product_id,
            'order_id': self.order_id,
            'quantity': self.quantity,
            'reason': self.reason,
            'status': self.status,
            'request_date': self.request_date.isoformat() if self.request_date else None,
            'processed_date': self.processed_date.isoformat() if self.processed_date else None,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }

File: frontend/mobile/lib/models/return_request.dart
class ReturnRequest {
  final int id;
  final int resellerId;
  final int productId;
  final int orderId;
  final int quantity;
  final String reason;
  final String status;
  final DateTime? requestDate;
  final DateTime? processedDate;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  ReturnRequest({
    required this.id,
    required this.resellerId,
    required this.productId,
    required this.orderId,
    required this.quantity,
    required this.reason,
    required this.status,
    this.requestDate,
    this.processedDate,
    this.createdAt,
    this.updatedAt,
  });

  factory ReturnRequest.fromJson(Map<String, dynamic> json) {
    return ReturnRequest(
      id: json['id'] ?? 0,
      resellerId: json['reseller_id'] ?? 0,
      productId: json['product_id'] ?? 0,
      orderId: json['order_id'] ?? 0,
      quantity: json['quantity'] ?? 0,
      reason: json['reason'] ?? '',
      status: json['status'] ?? 'pending',
      requestDate: json['request_date'] != null 
          ? DateTime.tryParse(json['request_date']) 
          : null,
      processedDate: json['processed_date'] != null 
          ? DateTime.tryParse(json['processed_date']) 
          : null,
      createdAt: json['created_at'] != null 
          ? DateTime.tryParse(json['created_at']) 
          : null,
      updatedAt: json['updated_at'] != null 
          ? DateTime.tryParse(json['updated_at']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'reseller_id': resellerId,
      'product_id': productId,
      'order_id': orderId,
      'quantity': quantity,
      'reason': reason,
      'status': status,
      'request_date': requestDate?.toIso8601String(),
      'processed_date': processedDate?.toIso8601String(),
      'created_at': createdAt?.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  // Getter untuk status color
  String get statusColor {
    switch (status.toLowerCase()) {
      case 'pending':
        return '#FF9800'; // Orange
      case 'approved':
        return '#4CAF50'; // Green
      case 'rejected':
        return '#F44336'; // Red
      case 'processed':
        return '#2196F3'; // Blue
      default:
        return '#9E9E9E'; // Gray
    }
  }

  // Getter untuk status display text
  String get statusDisplayText {
    switch (status.toLowerCase()) {
      case 'pending':
        return 'Menunggu';
      case 'approved':
        return 'Disetujui';
      case 'rejected':
        return 'Ditolak';
      case 'processed':
        return 'Diproses';
      default:
        return status;
    }
  }
}

class ReturnRequestCreate {
  final int productId;
  final int orderId;
  final int quantity;
  final String reason;

  ReturnRequestCreate({
    required this.productId,
    required this.orderId,
    required this.quantity,
    required this.reason,
  });

  Map<String, dynamic> toJson() {
    return {
      'product_id': productId,
      'order_id': orderId,
      'quantity': quantity,
      'reason': reason,
    };
  }
}

class OrderOption {
  final int orderId;
  final DateTime orderDate;
  final String status;
  final List<ProductOption> products;

  OrderOption({
    required this.orderId,
    required this.orderDate,
    required this.status,
    required this.products,
  });

  factory OrderOption.fromJson(Map<String, dynamic> json) {
    return OrderOption(
      orderId: json['id'],
      orderDate: DateTime.parse(json['order_date']),
      status: json['status'],
      products: (json['order_details'] as List)
          .map((e) => ProductOption.fromJson(e))
          .toList(),
    );
  }
}

class ProductOption {
  final int productId;
  final String productName;
  final int orderedQuantity;
  final int availableQuantity; // dari reseller stock

  ProductOption({
    required this.productId,
    required this.productName,
    required this.orderedQuantity,
    required this.availableQuantity,
  });

  factory ProductOption.fromJson(Map<String, dynamic> json) {
    return ProductOption(
      productId: json['product_id'],
      productName: json['product_name'] ?? 'Product ${json['product_id']}',
      orderedQuantity: json['quantity'],
      availableQuantity: json['available_quantity'] ?? 0,
    );
  }
}